internal/pkg/client/cache/dir.go

```diff
@@ -3,93 +3,308 @@
+	if c.rootDir == "" || c.rootDir == usr.HomeDir {
+		return false
+	}
+
+	// Basic check of the sub-cache validity
+	if c.Library == "" || c.Net == "" || c.OciTemp == "" || c.OciBlob == "" || c.Shub == "" {
+		return false
+	}
 
-	absdir, err := filepath.Abs(filepath.Join(root, subdir))
+	return true
+}
+
+// getCacheBaseDir figures out where the cache base directory is.
+func getCacheBasedir() (string, error) {
+	// If the user defined the special environment variable, we use its value
+	// as base directory.
+	basedir := os.Getenv(DirEnv)
+	if basedir != "" {
+		return basedir, nil
+	}
+
+	// If the environment variable is not set, we use the default cache.
+	usr, err := user.Current()
 	if err != nil {
-		sylog.Fatalf("Unable to get abs filepath: %v", err)
+		return "", fmt.Errorf("couldn't determine user home directory: %s", err)
+	}
+	basedir = path.Join(usr.HomeDir, basedirDefault)
+
+	return basedir, nil
+}
+
+// getCacheRoot figures out what the root directory is.
+func getCacheRoot(basedir string) string {
+	// Note: basedir and root are different and described earlier.
+	// The DirEnv environment variable can be used to set the basedir (or it
+	// is set to a default location) and the root is the actual cache directory
+	// within basedir, which we can safely delete (it is only supposed to contain
+	// files and directories that we create).
+	return path.Join(basedir, CacheDir)
+}
+
+// updateCacheSubdir update/create a sub-cache (directory) within the cache,
+// for example, the 'shub' cache.
+func updateCacheSubdir(c *SingularityCache, subdir string) (string, error) {
+	// This function may act on an cache object that is not fully initialized
+	// so it is not a method on a SingularityCache but rather an independent
+	// function.
+	// Because the cache object may not be initialized, we do NOT check its validity
+	if c == nil {
+		return "", fmt.Errorf("invalid cache handle")
+	}
+
+	// If the subdir is empty, it will lead to new collision since it would
+	// succeed but point at the cache's root
+	if subdir == "" {
+		return "", fmt.Errorf("invalid parameter")
+	}
+
+	absdir, err := filepath.Abs(filepath.Join(c.rootDir, subdir))
+	if err != nil {
+		return "", fmt.Errorf("unable to get abs filepath: %v", err)
 	}
 
 	if err := initCacheDir(absdir); err != nil {
-		sylog.Fatalf("Unable to initialize caching directory: %v", err)
+		return "", fmt.Errorf("unable to initialize caching directory: %v", err)
 	}
 
 	sylog.Debugf("Caching directory set to %s", absdir)
-	return absdir
+	return absdir, nil
 }
 
+// initCacheDir initializes a sub-cache within a cache, e.g., the shub
+// sub-cache.
 func initCacheDir(dir string) error {
-	if _, err := os.Stat(dir); os.IsNotExist(err) {
+	fInfo, err := os.Stat(dir)
+	switch {
+	case os.IsNotExist(err):
+		// The directory does not exist, we create it
 		sylog.Debugf("Creating cache directory: %s", dir)
 		if err := fs.MkdirAll(dir, 0755); err != nil {
```

Comment:
I'm sorry I didn't realize this before, but os.MkdirAll is already doing most of this: it begins by stat'ing dir to verify it exists, etc. Could you please check that we are not doing more work than necessary? I get the impression that simply calling fs.MkdirAll is enough.